### ISSUE TITLE
PEFT Training fails for dense models when PP = 1 when activation checkpointing is on

### ISSUE THREAD

**Issue by erictang000** [NVIDIA-NeMo/Megatron-Bridge#1750]
**Describe the bug**

<img width="358" height="231" alt="Image" src="https://github.com/user-attachments/assets/07fa7d00-1e42-40f7-9179-1b9ec67816e0" />

When integrating training with Megatron-Bridge PEFT with SkyRL, we're noticing issues with convergence for dense models that seem to be only happening with `pp = 1`. Upon digging further, I'm able to see that when `pp=1` for these dense models, the LoRA weights are not being updated after training, leading me to believe there's an issue in Megatron-Bridge or Megatron-Core code leading to this being the case. This also seems to be appearing on the verl integration: https://github.com/volcengine/verl/pull/4063#issuecomment-3661475306.

**Steps/Code to reproduce bug**

In this PR: https://github.com/NovaSky-AI/SkyRL/pull/743, running `skyrl-train/examples/megatron/run_megatron_lora_qwen3-0.6b.sh` with pp=1 will not converge, while pp > 1 will converge.

This might be too heavyweight for megatron-bridge to reproduce, but I'm wondering if there are PEFT training scripts (maybe for SFT) on the Megatron-Bridge end that might be able to reproduce this behavior by using pp=1 for dense models (like Qwen/Qwen3-0.6B). I wasn't able to find any recipes with those settings in this repo, so any pointers for repro'ing on Megatron-Bridge would be helpful too!

**Expected behavior**

I expect applying LoRA via Megatron-Bridge on a dense model with no model parallelism to correctly update weights.

**Additional context**

**Comment by HollowMan6** [comment]
I just found that for Verl's case, it will work (with non-zero grad norm) for PP=1 if we disable the activation checkpointing: https://github.com/volcengine/verl/pull/4063#issuecomment-3661475306 I'm not sure if this is also the case for you, but it looks like Megatron's CheckpointFunction.backward() is only invoked by PyTorch autograd when at least one input tensor requires grad.

- With PP>1, received tensors from other stages have requires_grad=True, so checkpoint backward is always called.
- With PP=1 and frozen base model, embedding outputs have requires_grad=False. This means CheckpointFunction.backward() is never called, and LoRA gradients inside the checkpoint are never computed.

**Comment by erictang000** [comment]
@HollowMan6 Thanks for the pointer! That worked for me as well. Seeing convergence with pp=1, tp=1, dp=4 now.

<img width="364" height="235" alt="Image" src="https://github.com/user-attachments/assets/e72798c5-a0a2-4e4b-9f59-401a9f650361" />

> With PP>1, received tensors from other stages have requires_grad=True, so checkpoint backward is always called.

I actually noticed while debugging that previously even with pp > 1 the lora updates on one of the ranks was 0 - I also reran with the same settings (pp=2, tp=1), but without activation checkpointing and found that the performance is better. Manually inspecting the lora updates, I can see the pp rank=0 worker was not updating.

<img width="1091" height="799" alt="Image" src="https://github.com/user-attachments/assets/a6bf235e-124f-42b7-ba78-8355d9d3bd9b" />

So it seems like we need a way to make sure that CheckpointFunction.backward() is invoked on the first pp stage in all cases

**Comment by yaoyu-33** [comment]
thanks @HollowMan6  @erictang000 . Updated the issue name to reflect the issue.

**Comment by HollowMan6** [comment]
I just opened https://github.com/volcengine/verl/pull/4566 to fix this on the verl side. Maybe similar fixes can be integrate into Megatron-LM, but for now I guess similar forward patching can be a viable solution for either skyRL or SFT for Megatron-Bridge with LoRA activation checkpointing.

**Comment by yaoyu-33** [comment]
I brought in @HollowMan6 's hook directly. https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1762/changes

**Comment by HollowMan6** [comment]
Ah yeah, it can be a better solution to patch just like #1762 so that it can benefit for all downstream projects. I will close https://github.com/volcengine/verl/pull/4566 and wait for #1762 to be merged. I will also test by integrating the #1762 patched Megatron-Bridge in Verl and will let you know if I find any issues.

**Comment by yaoyu-33** [comment]
@HollowMan6  thanks.

**Comment by yaoyu-33** [comment]
Closing as part of issue cleanup — this issue has been stale with no recent activity. If you still need help, please open a new issue with updated details. Thank you!

**Comment by pruprakash** [comment]
## QA RCCA Analysis

### 1. Fix Reference
- **PR**: #1762 - fix: Fix PEFT recompute hook
- **Issue**: #1750 - PEFT Training fails for dense models when PP=1 when activation checkpointing is on

### 2. Root Cause
When using PEFT (LoRA) with PP=1 and activation checkpointing enabled, the LoRA weights were not being updated after training. This was because the checkpointed transformer blocks did not see grad-requiring inputs during adapter-only finetuning, causing gradients to be dropped.

**Fix**: Added a reusable recompute helper to ensure checkpointed transformer blocks see grad-requiring inputs during adapter-only finetuning.

### 3. Trigger Configuration
- Dense model with LoRA/PEFT
- PP=1 (pipeline parallelism = 1)
- Activation checkpointing enabled (`recompute_granularity=full`)

### 4. Nature of the Bug
**Classification**: CODE BUG - Gradient flow broken with PEFT + recompute

### 5. Existing Test Coverage
**In Fix PR #1762**: YES
- `tests/unit_tests/peft/test_recompute.py` - unit test for grad-enable path and duplicate patch guard

**In NMFW Tests**: NO direct PEFT recompute tests

### 6. Coverage Assessment
| Test Type | Exists | Covers Bug |
|-----------|--------|------------|
| Unit test for PEFT recompute | YES (PR) | YES |

### 7. New Regression Test
**Not needed** - Fix PR #1762 includes unit test covering the grad-enable path.

### 8. Conclusion
**Verdict**: ADEQUATE COVERAGE
